lib/ansible/module_utils/facts/timeout.py

```diff
@@ -25,29 +26,30 @@
 DEFAULT_GATHER_TIMEOUT = 10
 
 
-class TimeoutError(Exception):
-    pass
+try:
+    # On Python 3, reuse the TimeoutError
+    class TimeoutError(TimeoutError):
+        pass
+except NameError:
+    # On Python 2, there unfortunately isn't a better base class to inherit from
+    class TimeoutError(Exception):
+        pass
 
 
 def timeout(seconds=None, error_message="Timer expired"):
-
     def decorator(func):
-        def _handle_timeout(signum, frame):
-            msg = 'Timer expired after %s seconds' % globals().get('GATHER_TIMEOUT')
-            raise TimeoutError(msg)
-
         def wrapper(*args, **kwargs):
-            local_seconds = seconds
-            if local_seconds is None:
-                local_seconds = globals().get('GATHER_TIMEOUT') or DEFAULT_GATHER_TIMEOUT
-            signal.signal(signal.SIGALRM, _handle_timeout)
-            signal.alarm(local_seconds)
+            timeout_value = seconds
+            if timeout_value is None:
+                timeout_value = globals().get('GATHER_TIMEOUT') or DEFAULT_GATHER_TIMEOUT
 
+            pool = mp.ThreadPool(processes=1)
+            res = pool.apply_async(func, args, kwargs)
+            pool.close()
             try:
-                result = func(*args, **kwargs)
-            finally:
-                signal.alarm(0)
-            return result
+                return res.get(timeout_value)
```

Comment:
I was testing this in my threading fact gathering branch,  the .get method won't help with 'blocked' processes and will inconsistently return or just hang forever (I'm still trying to get a reliable reproducer). I found this by 'simulating' bad NFS mounts by throttling the traffic to the nfs server the VMs are using.

Reply:
In py3 this is mitigated by the subproccess methods having a timeout, but this is not available on py2

Reply:
see my loop over waiting on threads here #49398 to get a py2/py3 compatible way

Reply:
We'll need a reproducer to be able to figure that out.  There's at least two other ways I can code this using threads but unless I can reproduce this I won't be able to work out what code is causing issues.

I tested this by running it on:
```
@timeout.timeout(1)
def slow_function():
    def run_command(['python', '-c', ' import time ; time.sleep(100)'])

try:
    slow_function()
except timeout.TimeoutError as e:
   print('Success: %s' % to_native(e))
```

and didn't see any problems so I'll need more to go on than that.

Reply:
its not a question of slow, but 'blocking' i.e accessing a bad NFS mount, that puts the process in 'B' status for the kernel and does not allow interruption.

Reply:
I'm unable to reproduce with https://gist.github.com/abadger/41ddf282503b94e84512dbbb7e9f93e0 but that's uninterruptible sleep state which is slightly different than blocking state.  If you can write a driver.sh for blocking state we can test that instead.

Reply:
Seems close enough, so I'll take that test as good as the 'unkillable' status is the main issue once the problem surfaces.

I still have not been able to reproduce the problem i saw reliably,  it just has happens a few time across the many times I've tested the gathering threaded code. It seems to happen less with Py3 versions, but since I'm not sure about how to trigger it, that is just anecdotal data. I was thinking of downgrading to nfsv3 since that was a lot more prone to this kind of issue ... but its probably not worth it.